src/particle/particle/particle.py

```diff
@@ -262,16 +256,14 @@ def table_names(cls):
             return tuple()
 
     @classmethod
-    def table_loaded(cls):
-        # type: () -> bool
+    def table_loaded(cls) -> bool:
         """
         Check to see if the table is loaded.
         """
         return cls._table is not None
 
     @classmethod
-    def all(cls):
-        # type: () -> List[Particle]
+    def all(cls: Type[Self]) -> List[Self]:
```

Comment:
Just for my education: here you have `cls: Type[Self]` but in several class methods above the same is not necessary (in `table_loaded`, for example). Why is that?

Reply:
You don't need to annotate the first argument (self or cls); the reason we do it is to pin a TypeVar. TypeVar's _must_ be added to a "known" type before they can be used to describe a type. Since MyPy knows `cls`'s type, it can use that to inver the return type `List[Self]`. If `cls` is `Particle`, `Self` is an instance of Particle. If `cls` is a subclass of `Particle`, then `Self` is that subclass.

If [PEP 673](https://www.python.org/dev/peps/pep-0673/) is accepted, then we can do `from typing/typing_extensions import Self`, then drop `Self/Type[Self]` on self/cls and _just_ use it for returns, which would be _much_ nicer than all these TypeVar's. :) PEP 673 was submitted to the Python SC yesterday! :)